### ISSUE TITLE
Project modernization

### ISSUE THREAD

**Issue by mbugni** [jotd666/gods-deluxe#4]
#### Once upon a time
I loved this game when I was student, during an ancient time.

#### Today
I'm a Java developer and a Linux expert user, not a game developer.

## Goal
I [forked this project](https://github.com/mbugni/gods-deluxe) to experiment with some enhancement, especially to modernize code, building and dependencies.

#### Build enhancements
I used Maven to automate building and create a single `.jar` executable.
An optional `.zip` containing just assets can be done as well.

The game can be run now with something like:
```shell
java -DROOT_DIR="/your/assets/folder" -jar gods-deluxe.jar
```

I adopted [Maven Wrapper](https://maven.apache.org/wrapper/) for portable builds:
```shell
./mvnw clean package 
```

#### Code enhancements (to do)
- Minimum requirements: Java 8 (or later) and 64 bits OS (Windows/Linux).
- Replace obsolete dependencies with more recent ones.

#### Delivery  (to do)
Automatic builds will make simpler to create self-contained deliverables. My plan is to provide a [Flatpak app](https://flatpak.org/) for Linux users and possibly publish it on [Flathub](https://flathub.org/), just like I did with other games.

I think something similar can be done for Windows, but I'm not an expert of that platform.

## Conclusion
@jotd666  what do you think about?

**Comment by mbugni** [comment]
Meanwhile I opened a [pull request](https://github.com/jotd666/gods-deluxe/pull/5).

**Comment by jotd666** [comment]
Hi,

Thanks for this. Glad that someone takes interest in the project. Doing a .jar instead of all those .classes is certainly a good idea. Is it possible to also put the resources (non-code) in the .jar file to get rid of ROOT_DIR variable ?

**Comment by mbugni** [comment]
Thanks for accepting my collaboration.

> Is it possible to also put the resources (non-code) in the .jar file to get rid of ROOT_DIR variable ?

It's one of the goal I would like to get. The next release could contain 3 artifacts (automatically generated by Maven):
- `gods-deluxe-engine.jar` -> Engine only
- `gods-deluxe-assets.zip` -> Immutable assets only
- `gods-deluxe.jar` -> Engine and assets

For those who wants to experiments with assets (like sounds/music):
```shell
java -DROOT_DIR="/your/assets/folder" -jar gods-deluxe-engine.jar
```

For those who want just play:
```shell
java -jar gods-deluxe.jar
```

But, if resources are well managed, it could be just one `.jar` (engine + assets) that can recognize the `ROOT_DIR`
parameter and use it (if provided). Probably the best solution.

Before to do that, I suggest to improve other areas like:
- Full Java 8 porting (JVM 1.8 it's available even for old 32-bit systems)
- Relocable configurations (eg using `.properties`)
- Remove local solutions in favour of Java APIs

If you agree, we can open issues for each activity and Git branches to manage them as PR.

**Comment by mbugni** [comment]
> Is the PR thing necessary now that you have access to the repository?

No, but four eyes are better than two. There are thing that I'm not able to test, like the editor.
If you can take a look to PRs, it will be appreciated.

> Another issue: my old webpage is now gone, and there's no binary distribution of the game or even the music packs.

Do you mean [this one](https://jotd.pagesperso-orange.fr/gods/) !?

I mirrored that, just in case.

I can turn the `readme.txt` into markdown format, so it can be the landing page for the project.
 
> Those should be in the "RELEASE" section of github maybe.

Yes, GitHub releases area is the right place for deliverables.

> I'm not paying hosting for that.

You can use [GitHub pages](https://pages.github.com/) for free. This is [my page](https://mbugni.github.io/).

**Comment by jotd666** [comment]
My old website is down now.

Thanks for your hard work!

**Comment by mbugni** [comment]
### Current status (0.9.0-SNAPSHOT)
- Filesystem issues are fixed (file names, paths, etc.) for Linux (MacOS user should benefit too)
- Old/embedded dependencies replaced with most recently available
- Assets and data folders can be relocated (by providing optional parameters `-DGODS_ASSETS_DIR` and `-DGODS_DATA_DIR`)
- Classes and dependencies are packed in a single `.jar`
- Metadata for Linux packaging are provided
- The distro zip file can be created with command `./mvnw package`

I think a release can be done.

**Comment by jotd666** [comment]
Awesome work. thanks for the update!



Le jeu. 26 oct. 2023, 12:54, Massi aka Ergosum ***@***.***> a
écrit :

**Comment by mbugni** [comment]
@jotd666 can you please take a look at [gods-deluxe.metainfo.xml](https://github.com/jotd666/gods-deluxe/blob/97687a0c187b5a533095b7d34a3d76ab31ce803d/delivery/metadata/gods-deluxe.metainfo.xml) and check if everything is ok for you (author name, description etc.)?

I reported some info from the old website.

**Comment by mbugni** [comment]
### Publishing on Flathub (for Linux users)

Now Gods Deluxe can be installed via [Flatpak](https://flatpak.org/) from this link:

[![Download Doom Runner](https://dl.flathub.org/assets/badges/flathub-badge-en.png 'Download Gods Deluxe')](https://flathub.org/apps/io.github.jotd666.gods-deluxe)

**Comment by mbugni** [comment]
### Release 0.10.0
This new release brings new features:
-  [Resizable game window](https://github.com/jotd666/gods-deluxe/pull/11) with[ antialiasing support](https://github.com/jotd666/gods-deluxe/pull/13)
- [Sound optimization](https://github.com/jotd666/gods-deluxe/pull/12) for less resources usage and faster start up
- Improved portability/[default settings](https://github.com/jotd666/gods-deluxe/pull/14)

**Comment by jotd666** [comment]
just had a quick run. I love the resizable window. It's great!

**Comment by mbugni** [comment]
I'm glad to read this. I think it's a good improvement for modern displays.
